server_test.go

```diff
@@ -444,3 +444,68 @@ func TestDecodeContext(t *testing.T) {
 		testDecodeContext(t, h1, h2, h1.ID())
 	})
 }
+
+func TestAuthorization(t *testing.T) {
+	h1, h2 := makeRandomNodes()
+	defer h1.Close()
+	defer h2.Close()
+
+	h3, _ := libp2p.New(
+		context.Background(),
+		libp2p.ListenAddrStrings("/ip4/127.0.0.1/tcp/19997"),
+	)
+	h3.Peerstore().AddAddrs(h1.ID(), h1.Addrs(), peerstore.PermanentAddrTTL)
+
+	authorizationFunc := AuthorizeWithMap(
+		map[peer.ID]map[string]bool{
+			h2.ID(): map[string]bool{
+				"Arith.Multiply": true,
+			},
+		},
+	)
+
+	s := NewServer(h1, "rpc", WithAuthorizeFunc(authorizationFunc))
+	c := NewClientWithServer(h2, "rpc", s)
+	var arith Arith
+	s.Register(&arith)
+
+	dest := h1.ID()
+
+	var r int
+	err := c.Call(dest, "Arith", "Multiply", &Args{2, 3}, &r)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if r != 6 {
+		t.Error("result is:", r)
+	}
+
+	var q Quotient
+	err = c.Call(dest, "Arith", "Divide", &Args{20, 6}, &q)
+	if err == nil {
+		t.Fatal("expected error instead")
+	}
+	expectedErrMsg := "client does not have permissions to this method, service name: Arith, method name: Divide"
+	actualErrMsg := err.Error()
+	if actualErrMsg != expectedErrMsg {
```

Comment:
I think we need to extend the `responseErr` enum to include `authorizationError` for the following reasons:
1. String comparison/regex to determine how a method failed is fragile, it means no one can change that error message without break all tests and downstream code
2. It means that clients of `gorpc` can't **reliably** make programmatic choices if there is an authorization error, i.e. without inspecting the error string how do I tell whether the error is a network error or an auth error

Ideally, in the long run `errors.go` moves towards a more standard code-based system but for this I think just adding another variant of serverError/clientError will serve the purpose sufficiently.

Reply:
Addressed with latest commits